src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1420 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.provider;
+
+import jdk.internal.vm.annotation.IntrinsicCandidate;
+import sun.security.provider.SHA3.SHAKE128;
+import sun.security.provider.SHA3.SHAKE256;
+
+import java.util.Arrays;
+
+public class ML_DSA {
+    private static final int mlDsa_q = 8380417;
+    private static final int mlDsa_n = 256;
+    private static final int shake256BlockSize = 136; // the block length for SHAKE256
+    private static final int montRBits = 32;
+    private static final long montR = 4294967296L; // 1 << montRBits
+    private static final int montQ = 8380417;
+    private static final int montRSquareModQ = 2365951;
+    private static final int montQInvModR = 58728449;
+    private static final int montRModQ = 4193792;
+    private static final int montDimInverse = 16382; // toMont((mlDsa_n)^-1 (mod mlDsa_q))
+    private static final int[] montZetasForNtt = new int[]{
```

Comment:
Some of these values are constants in the spec and others are commonly used values derived from the constants. For example, `s1PackedLength = (mlDsa_l * (eta + 1) * 256) / 8` because the vector `s1` contains `mlDsa_l` polynomials of length 256 with 3-bit coefficients (the final division by 8 is to represent it in bytes). Perhaps it would be useful to annotate these values with how they're computed?